### PR TITLE
ROCm support

### PR DESCRIPTION
**Pull request by chhwang**

### CODE REVIEW

File: include/mscclpp/device.hpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#define MSCCLPP_ON_DEVICE
+#define MSCCLPP_ON_HOST_DEVICE
+
+#if defined(__CUDA_ARCH__)
+#define MSCCLPP_CUDA
+#define MSCCLPP_DEVICE_INLINE __forceinline__ __device__
+#elif defined(__HIP_DEVICE_COMPILE__)
+#include <hip/hip_runtime.h>
+#define MSCCLPP_HIP
+#define MSCCLPP_DEVICE_INLINE __device__ inline
```
Comment (chhwang) [inline include/mscclpp/device.hpp:20]: Need to check if we can just use `__forceinline__`

File: include/mscclpp/errors.hpp [hunk 2]
```diff
@@ -77,14 +74,14 @@ class SysError : public BaseError {
 /// An error from a CUDA runtime library call.
 class CudaError : public BaseError {
  public:
-  CudaError(const std::string& message, cudaError_t errorCode);
+  CudaError(const std::string& message, int errorCode);
   virtual ~CudaError() = default;
 };
 
 /// An error from a CUDA driver library call.
 class CuError : public BaseError {
```
Comment (chhwang) [inline include/mscclpp/errors.hpp:82]: `CuError` is unused now

File: include/mscclpp/fifo_device.hpp [hunk 3]
```diff
@@ -16,8 +17,18 @@ namespace mscclpp {
...
+  struct Data {
+    uint64_t fst;
+    uint64_t snd;
+  } data;
+  struct {
+    uint64_t fst;
+    uint64_t snd;
+  };
+#if defined(MSCCLPP_ON_HOST_DEVICE)
+  longlong2 raw_;
```
Comment (chhwang) [inline include/mscclpp/fifo_device.hpp:30]: Use casting instead of a union member?

File: include/mscclpp/packet.hpp [hunk 4]
```diff
@@ -18,67 +19,74 @@ union LLPacket {
...
-    asm volatile("st.volatile.global.v4.u32 [%0], {%1,%2,%3,%4};" ::"l"(v), "r"(val1), "r"(flag), "r"(val2), "r"(flag));
+  MSCCLPP_DEVICE_INLINE void write(uint32_t val1, uint32_t val2, uint32_t flag) {
+#if defined(MSCCLPP_CUDA) || defined(MSCCLPP_CUDA_HOST)
+    asm volatile("st.volatile.global.v4.u32 [%0], {%1,%2,%3,%4};" ::"l"(&raw_), "r"(val1), "r"(flag), "r"(val2),
+                 "r"(flag));
+#else
+    uint4 reg = make_uint4(val1, flag, val2, flag);
+    ulonglong2* p = reinterpret_cast<ulonglong2*>(&reg);
+    atomicStore(&(raw_.x), p->x, memoryOrderRelaxed);
+    atomicStore(&(raw_.y), p->y, memoryOrderRelaxed);
```
Comment (chhwang) [inline include/mscclpp/packet.hpp:39]: Use `__builtin_nontemporal_store` instead

File: include/mscclpp/packet.hpp [hunk 5]
```diff
@@ -18,67 +19,74 @@ union LLPacket {
...
     uint32_t flag1, flag2;
     asm volatile("ld.volatile.global.v4.u32 {%0,%1,%2,%3}, [%4];"
                  : "=r"(data.x), "=r"(flag1), "=r"(data.y), "=r"(flag2)
-                 : "l"(v));
+                 : "l"(&raw_));
     return (flag1 != flag) || (flag2 != flag);
+#else
+    ulonglong2 reg;
+    reg.x = atomicLoad(&(raw_.x), memoryOrderRelaxed);
+    reg.y = atomicLoad(&(raw_.y), memoryOrderRelaxed);
```
Comment (chhwang) [inline include/mscclpp/packet.hpp:62]: Use `__builtin_nontemporal_load` instead

**Review by yzygitzh — APPROVED** [review]

File: include/mscclpp/device.hpp (hunk 1 above)
Comment (chhwang) [reply include/mscclpp/device.hpp]: Tackled

File: include/mscclpp/fifo_device.hpp (hunk 3 above)
Comment (chhwang) [reply include/mscclpp/fifo_device.hpp]: Tackled

File: include/mscclpp/packet.hpp (hunk 4 above)
Comment (chhwang) [reply include/mscclpp/packet.hpp:44]: Adding a TODO

File: include/mscclpp/packet.hpp (hunk 5 above)
Comment (chhwang) [reply include/mscclpp/packet.hpp:68]: Adding a TODO

File: test/mp_unit/proxy_channel_tests.cu [hunk 6]
```diff
@@ -362,7 +362,7 @@ void ProxyChannelOneToOneTest::testPacketPingPongPerf(bool useIbOnly) {
 
   auto* testInfo = ::testing::UnitTest::GetInstance()->current_test_info();
   const std::string testName = std::string(testInfo->test_suite_name()) + "." + std::string(testInfo->name());
-  const int nTries = 1000;
+  const int nTries = 1000000;
```
Comment (cp5555) [inline test/mp_unit/proxy_channel_tests.cu:365]: What's the purpose to enlarge this?

File: test/mp_unit/proxy_channel_tests.cu [hunk 7]
```diff
@@ -242,14 +242,14 @@ __global__ void kernelProxyLLPingPong(int* buff, mscclpp::LLPacket* putPktBuf, m
       if (CheckCorrectness) {
         // If each thread reads 8 bytes at once, we don't need a barrier after getPackets().
         // __syncthreads();
-        for (int j = threadId; j < nPkt; j += numThreads) {
-          if (buffPtr[2 * j] != getOffset + i + 2 * j) {
+        for (size_t j = threadId; j < nPkt; j += numThreads) {
+          if (buffPtr[2 * j] != getOffset + i + 2 * (int)j) {
```
Comment (cp5555) [inline test/mp_unit/proxy_channel_tests.cu:246]: j is size_t, but here is int. It's strange.

File: test/mp_unit/ib_tests.cu [hunk 8]
```diff
@@ -118,6 +118,8 @@ __global__ void kernelMemoryConsistency(uint64_t* data, volatile uint64_t* curIt
                                         uint64_t nelem, uint64_t maxIter) {
   if (blockIdx.x != 0) return;
 
+  __shared__ int errs[1024];
```
Comment (cp5555) [inline test/mp_unit/ib_tests.cu:121]: How can you guarantee it doesn't exceed 1024?

File: cmake/check_amd_gpu.hip [hunk 9]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline cmake/check_amd_gpu.hip:1]: What's goal for this?

File: python/mscclpp/error_py.cpp [hunk 10]
```diff
@@ -31,10 +31,10 @@ void register_error(nb::module_& m) {
       .def(nb::init<const std::string&, int>(), nb::arg("message"), nb::arg("errorCode"));
 
   nb::class_<CudaError, BaseError>(m, "CudaError")
-      .def(nb::init<const std::string&, cudaError_t>(), nb::arg("message"), nb::arg("errorCode"));
+      .def(nb::init<const std::string&, int>(), nb::arg("message"), nb::arg("errorCode"));
```
Comment (cp5555) [inline python/mscclpp/error_py.cpp:34]: cudaError_t to int?

**Review by cp5555 — APPROVED** [review]

File: test/mp_unit/proxy_channel_tests.cu (hunk 6 above)
Comment (chhwang) [reply test/mp_unit/proxy_channel_tests.cu:365]: I enlarged to make numbers more stable. Actually it was an issue only for H100 so far

File: test/mp_unit/ib_tests.cu (hunk 8 above)
Comment (chhwang) [reply test/mp_unit/ib_tests.cu:121]: This test uses only one thread block that can have up to 1024 threads.

File: cmake/check_amd_gpu.hip (hunk 9 above)
Comment (chhwang) [reply cmake/check_amd_gpu.hip:1]: We compile & run this code during cmake build to check if an AMD GPU actually exists on the system. Only checking cuda/rocm installation is not enough as we sometimes install both during dev (for hipify-clang).

File: python/mscclpp/error_py.cpp (hunk 10 above)
Comment (chhwang) [reply python/mscclpp/error_py.cpp:34]: I changed this interface to get an `int` input and cast it into `cudaError_t` internally.

File: test/allgather_test_cpp.cu [hunk 11]
```diff
@@ -74,7 +74,12 @@ __device__ void localAllGather(DeviceHandle<mscclpp::SimpleProxyChannel> proxyCh
     if ((remoteRank % nranksPerNode) == ((rank - i + nranksPerNode) % nranksPerNode)) {
       if ((threadIdx.x % 32) == 0) proxyChan.wait();
     }
+#if defined(__HIP_PLATFORM_AMD__)
+    // TODO: group barrier
```
[the remark this answers is not in the document]
Comment (chhwang) [reply test/allgather_test_cpp.cu]: I fixed the comment. We actually need a group barrier here for better performance, but `__syncthreads()` is still correct. This code is not for perf test and perhaps will be deprecated in the future.

File: test/mp_unit/proxy_channel_tests.cu (hunk 7 above)
Comment (chhwang) [reply test/mp_unit/proxy_channel_tests.cu]: Thanks for the comment, I fixed this.

**Pull request merged**